### PR TITLE
fix(router): Preloader should use the module injector from a route

### LINKED ISSUE #15626: CanDeactivate being called with null component and providers not being provided to new component
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
After upgrading working app from 4.0.0 to 4.0.1 or 4.1.0-beta.0, when navigating to a new component (lazy loaded but dont know if that matters yet) CanDeactivate guard is called with null component and providers are not being provided to new component (sometimes I get only one of these errors, sometimes I get both as shown below, and occasionally I get other errors not captured here)

```
ERROR Error: Uncaught (in promise): Error: No provider for MyService!
Error
    at injectionError (core.es5.js:1231) [angular]
    at noProviderError (core.es5.js:1269) [angular]
    at ReflectiveInjector_._throwOrNull (core.es5.js:2770) [angular]
    at ReflectiveInjector_._getByKeyDefault (core.es5.js:2809) [angular]
    at ReflectiveInjector_._getByKey (core.es5.js:2741) [angular]
    at ReflectiveInjector_.get (core.es5.js:2610) [angular]
    at AppModuleInjector.NgModuleInjector.get (core.es5.js:3557) [angular]
    at MySearchModuleInjector.NgModuleInjector.get (core.es5.js:3557) [angular]
    at resolveDep (core.es5.js:10930) [angular]
    at createClass (core.es5.js:10799) [angular]
    at createDirectiveInstance (core.es5.js:10627) [angular]
    at createViewNodes (core.es5.js:11977) [angular]
    at createRootView (core.es5.js:11882) [angular]
    at callWithDebugContext (core.es5.js:13013) [angular]
    at injectionError (core.es5.js:1231) [angular]
    at noProviderError (core.es5.js:1269) [angular]
    at ReflectiveInjector_._throwOrNull (core.es5.js:2770) [angular]
    at ReflectiveInjector_._getByKeyDefault (core.es5.js:2809) [angular]
    at ReflectiveInjector_._getByKey (core.es5.js:2741) [angular]
    at ReflectiveInjector_.get (core.es5.js:2610) [angular]
    at AppModuleInjector.NgModuleInjector.get (core.es5.js:3557) [angular]
    at MySearchModuleInjector.NgModuleInjector.get (core.es5.js:3557) [angular]
    at resolveDep (core.es5.js:10930) [angular]
    at createClass (core.es5.js:10799) [angular]
    at createDirectiveInstance (core.es5.js:10627) [angular]
    at createViewNodes (core.es5.js:11977) [angular]
    at createRootView (core.es5.js:11882) [angular]
    at callWithDebugContext (core.es5.js:13013) [angular]
    at resolvePromise (zone.js:712) [angular]
    at resolvePromise (zone.js:683) [angular]
    at :4200/polyfills.bundle.js:2764:17 [angular]
    at Object.onInvokeTask (core.es5.js:4116) [angular]
    at ZoneDelegate.webpackJsonp.626.ZoneDelegate.invokeTask (zone.js:397) [angular]
    at Zone.webpackJsonp.626.Zone.runTask (zone.js:165) [<root> => angular]
    at drainMicroTaskQueue (zone.js:593) [<root>]
    at HTMLAnchorElement.ZoneTask.invoke (zone.js:464) [<root>]
defaultErrorLogger @ core.es5.js:1084
ErrorHandler.handleError @ core.es5.js:1144
next @ core.es5.js:4754
schedulerFn @ core.es5.js:3828
SafeSubscriber.__tryOrUnsub @ Subscriber.js:234
SafeSubscriber.next @ Subscriber.js:183
Subscriber._next @ Subscriber.js:125
Subscriber.next @ Subscriber.js:89
Subject.next @ Subject.js:55
EventEmitter.emit @ core.es5.js:3814
NgZone.triggerError @ core.es5.js:4185
onHandleError @ core.es5.js:4146
webpackJsonp.626.ZoneDelegate.handleError @ zone.js:369
webpackJsonp.626.Zone.runGuarded @ zone.js:141
_loop_1 @ zone.js:604
drainMicroTaskQueue @ zone.js:613
ZoneTask.invoke @ zone.js:464
core.es5.js:1084 

ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'canDeactivate' of null
TypeError: Cannot read property 'canDeactivate' of null
    at CanDeactivateGuard.webpackJsonp.224.CanDeactivateGuard.canDeactivate (CanDeactivate.ts:9)
    at MergeMapSubscriber.project (router.es5.js:4394)
    at MergeMapSubscriber._tryNext (mergeMap.js:120)
    at MergeMapSubscriber._next (mergeMap.js:110)
    at MergeMapSubscriber.Subscriber.next (Subscriber.js:89)
    at ArrayObservable._subscribe (ArrayObservable.js:114)
    at ArrayObservable.Observable._trySubscribe (Observable.js:57)
    at ArrayObservable.Observable.subscribe (Observable.js:45)
    at MergeMapOperator.call (mergeMap.js:85)
    at Observable.subscribe (Observable.js:42)
    at EveryOperator.call (every.js:33)
    at Observable.subscribe (Observable.js:42)
    at Object.subscribeToResult (subscribeToResult.js:22)
    at MergeMapSubscriber._innerSub (mergeMap.js:130)
    at MergeMapSubscriber._tryNext (mergeMap.js:127)
    at CanDeactivateGuard.webpackJsonp.224.CanDeactivateGuard.canDeactivate (CanDeactivate.ts:9)
    at MergeMapSubscriber.project (router.es5.js:4394)
    at MergeMapSubscriber._tryNext (mergeMap.js:120)
    at MergeMapSubscriber._next (mergeMap.js:110)
    at MergeMapSubscriber.Subscriber.next (Subscriber.js:89)
    at ArrayObservable._subscribe (ArrayObservable.js:114)
    at ArrayObservable.Observable._trySubscribe (Observable.js:57)
    at ArrayObservable.Observable.subscribe (Observable.js:45)
    at MergeMapOperator.call (mergeMap.js:85)
    at Observable.subscribe (Observable.js:42)
    at EveryOperator.call (every.js:33)
    at Observable.subscribe (Observable.js:42)
    at Object.subscribeToResult (subscribeToResult.js:22)
    at MergeMapSubscriber._innerSub (mergeMap.js:130)
    at MergeMapSubscriber._tryNext (mergeMap.js:127)
    at resolvePromise (zone.js:712) [angular]
    at resolvePromise (zone.js:683) [angular]
    at :4200/polyfills.bundle.js:2764:17 [angular]
    at Object.onInvokeTask (core.es5.js:4116) [angular]
    at ZoneDelegate.webpackJsonp.626.ZoneDelegate.invokeTask (zone.js:397) [angular]
    at Zone.webpackJsonp.626.Zone.runTask (zone.js:165) [<root> => angular]
    at drainMicroTaskQueue (zone.js:593) [<root>]
    at HTMLAnchorElement.ZoneTask.invoke (zone.js:464) [<root>]
```

**Expected behavior**
CanDeactivate guard should not be called with null component and providers should be provided

**Minimal reproduction of the problem with instructions**

Will try to create plnkr when have time.

**What is the motivation / use case for changing the behavior?**

Breaking change. Causes crash in existing code that works as expected in 4.0.0

**Please tell us about your environment:**

@angular/cli: 1.0.0
node: 7.7.1
os: win32 x64
@angular/common: 4.1.0-beta.0
@angular/compiler: 4.1.0-beta.0
@angular/core: 4.1.0-beta.0
@angular/forms: 4.1.0-beta.0
@angular/http: 4.1.0-beta.0
@angular/platform-browser: 4.1.0-beta.0
@angular/platform-browser-dynamic: 4.1.0-beta.0
@angular/router: 4.1.0-beta.0
@angular/cli: 1.0.0
@angular/compiler-cli: 4.1.0-beta.0

* **Angular version:** 2.0.X

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 Only tried in Chrome

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]

* **Node (for AoT issues):** `node --version` =

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes regression introduced by https://github.com/angular/angular/commit/23bf34853c83b2131991e11da1c8dcc75ee3d731#diff-86faca9d9f6af67bf4abf6c3f6ab380bR108

Closes #15626

### CODE REVIEW

**Comment by DzmitryShylovich** [comment]
@vicb I didn't add a test because the repro from the original issue is too big and I don't have time to figure out how to write a small test case. Could you take over this pr and add a test?

File: packages/router/src/config.ts [hunk 1]
```diff
@@ -363,7 +364,7 @@ export interface Route {
 
 export interface InternalRoute extends Route {
   // `LoadedRouterConfig` loaded via a Route `loadChildren`
-  _loadedConfig?: any;
+  _loadedConfig?: LoadedRouterConfig|undefined;
```
Comment (vicb) [inline packages/router/src/config.ts]: - the `?:` means `|undefined`
- there is not type here to avoid a circular ref

**Comment by vicb** [comment]
Great catch @DzmitryShylovich, I'll cherry pick the second commit and add a test.

Could you please create an other PR out of the first commit ? 

Thanks

**Comment by DzmitryShylovich** [comment]
> Could you please create an other PR out of the first commit ?

ok will do.

> there is not type here to avoid a circular ref

I can move `LoadedRouterConfig` into another file to avoid a circular ref

**Comment by DzmitryShylovich** [comment]
new pr https://github.com/angular/angular/pull/15775

**Pull request closed without merging**